### PR TITLE
Import dune-mode from tuareg's repository

### PR DESCRIPTION
**Pull request by rgrinberg**
I checked the original repo, and the only contributors to this file were me and @Chris00. I believe we both have re-assigned our copyright to janestreet at that time, so we can just import this file without problems.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Btw. For modularity, I think we should move the linting stuff to `dune-flymake`. In my opinion, dune should support linting via lsp so that it works for all editors and not just emacs.

File: editor-integration/emacs/dune-mode.el [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  nil
+  "(install" > \n
+  "((section " _ ")" > \n
+  "(files (" _ "))))" > ?\n)
+
+(define-skeleton dune-insert-copyfiles-form
+  "Insert a copy_files stanza."
+  nil
+  "(copy_files " _ ")" > ?\n)
+
```
Comment (Chris00) [inline editor-integration/emacs/dune-mode.el]: I'll add some new skeletons once this is merged.

File: editor-integration/emacs/dune-mode.el [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    (define-key map "\C-c.i" 'dune-insert-install-form)
+    (define-key map "\C-c.c" 'dune-insert-copyfiles-form)
+    map)
+  "Keymap used in dune mode.")
+
+(defun dune-build-menu ()
+  (easy-menu-define
+    dune-mode-menu  (list dune-mode-map)
+    "dune mode menu."
+    '("Dune/jbuild"
```
Comment (Chris00) [inline editor-integration/emacs/dune-mode.el]: `Dune/jbuild` → `Dune`

File: editor-integration/emacs/dune-mode.el [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+               "run" "chdir" "setenv"
+               "with-stdout-to" "with-stderr-to" "with-outputs-to"
+               "ignore-stdout" "ignore-stderr" "ignore-outputs"
+               "progn" "echo" "write-file" "cat" "copy" "copy#" "system"
+               "bash" "diff" "diff?"
+               ;; inline_tests and inline_tests.backend
+               ;; FIXME: "flags" is already a field and we do not have enough
+               ;; context to distinguishing both.
+               "backend" "generate_runner" "runner_libraries" "flags"
+               "extends")
```
Comment (Chris00) [inline editor-integration/emacs/dune-mode.el]: This is not up-to-date with the current Dune.  I'll do a PR with my WIP once this is merged but some help is welcome.

File: editor-integration/emacs/dune-flymake.el [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+(require 'flymake)
+(require 'dune-mode)
+
+(defvar dune-flymake-temporary-file-directory
+  (expand-file-name "dune" temporary-file-directory)
+  "Directory where to duplicate the files for flymake.")
+
+(defvar dune--allowed-file-name-masks
+  '("\\(?:\\`\\|/\\)jbuild\\'" dune-flymake-init
```
Comment (Chris00) [inline editor-integration/emacs/dune-flymake.el]: jbuild → dune
(WIP in my copy)

File: editor-integration/emacs/dune-flymake.el [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    dune-flymake-cleanup)
+  "Flymake entry for dune files.  See `flymake-allowed-file-name-masks'.")
+
+(defvar dune--err-line-patterns
+  ;; Beware that the path from the root will be reported by dune
+  ;; but flymake requires it to match the file name.
+  '(("File \"[^\"]*\\(jbuild\\)\", line \\([0-9]+\\), \
```
Comment (Chris00) [inline editor-integration/emacs/dune-flymake.el]: jbuild → dune

File: editor-integration/emacs/dune-flymake.el [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  (let ((fname (flymake-init-create-temp-buffer-copy
+                'dune-flymake-create-temp))
+        (root (or (dune--root buffer-file-name) "")))
+    (list dune-program (list fname root))))
+
+(defun dune-flymake-dune-mode-hook ()
+  (push dune--allowed-file-name-masks flymake-allowed-file-name-masks)
+  (setq-local flymake-err-line-patterns dune--err-line-patterns))
+
+(add-hook )
```
Comment (Chris00) [inline editor-integration/emacs/dune-flymake.el]: ?

File: editor-integration/emacs/dune-mode.el [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+;; LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+;; NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+;; CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+(require 'scheme)
+
+(defvar dune-mode-hook nil
+  "Hooks for the `dune-mode'.")
+
+(defvar dune-program
```
Comment (Chris00) [inline editor-integration/emacs/dune-mode.el]: I think it is better declared in `dune-flymake`.

File: editor-integration/emacs/dune-mode.el [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+       (dune--field-vals "kind" "normal" "ppx_rewriter" "ppx_deriver"))
+    ,(eval-when-compile
+       (dune--field-vals "mode" "standard" "fallback" "promote"
+                                "promote-until-clean"))
+    (,(concat "(" dune-builtin-regex) 1 font-lock-builtin-face)
+    ("(preprocess[[:space:]]+(\\(pps\\)" 1 font-lock-builtin-face)
+    (,(eval-when-compile
+        (concat "(" (regexp-opt '("fallback") t)))
+     1 dune-error-face)
+    (,(concat "${" dune-var-regex "}")
```
Comment (Chris00) [inline editor-integration/emacs/dune-mode.el]: This is now incorrect — unpublished WIP in my copy.

**Comment by diml** [comment]
BTW, for the skeletons, keywords, etc... all this could be generated by dune. We have a background WIP to make the parser be a pure applicative, currently it's a monad. Once this is done it will be trivial to extract this information

File: editor-integration/emacs/dune-mode.el (hunk 8 above)
Comment (rgrinberg) [reply editor-integration/emacs/dune-mode.el]: What should I do about this?

**Comment by rgrinberg** [comment]
@Chris00 addressed your comments except the one I don't know what to do about. Feel free to push your fixes here.

**Comment by Chris00** [comment]
See https://github.com/ocaml/dune/pull/1265

**Comment by Chris00** [comment]
> Feel free to push your fixes here.

Not sure how to do that when I'm not the owner of the PR.

**Comment by rgrinberg** [comment]
> Not sure how to do that when I'm not the owner of the PR.

You can just my git repo as a remote, and push to my origin. Anyways, creating your own PR is fine too.

**Comment by Chris00** [comment]
>  Once this is done it will be trivial to extract this information

Cool.

**Pull request closed without merging**